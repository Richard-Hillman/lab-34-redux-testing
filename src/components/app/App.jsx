// /* eslint-disable max-len */
import React, { useReducer } from 'react';


const initialState = {
  color: 'red',
  before: [],
  after: []
};

function reducer(state, action) {
  switch(action.type) {
    case 'COLOR_CHANGE': {
      const before = [...state.before, state.color];
      return { color: action.payload, before };
    }
  
    
    // --------------------------------------------

    case 'COLOR_UNDO':
      return { color: action.payload };

      // --------------------------------------------

    case 'COLOR_REDO':
      return { color: action.payload };

  }
}
// --------------------------------------------

function App() {
  const [state, dispatch] = useReducer(reducer, initialState);

  const record = ({ target }) => {

    dispatch({
      type: (target.id),
      payload: target.value
    });
  };

  return (
    <>

      {/* ---------------------------------------------------- */}

    
      {/* <button onClick={ undo }>undo</button>

      <button onClick={redo}>redo</button>
      */}
      <label htmlFor="COLOR_CHANGE">
        Current Color
      </label>
      
      {/* ---------------------------------------------------- */}


      <input 
        id="COLOR_CHANGE" 
        type="color"
        value={state.color}
        onChange={record}      
      /> 

      {/* ---------------------------------------------------- */}

      <div data-testid="frig"
        style={{
          backgroundColor: state.color,
          width: '10rem',
          height: '10rem' }}>
      </div>
    </>
  );
}

export default App;
