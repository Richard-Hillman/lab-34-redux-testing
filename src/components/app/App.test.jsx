/* eslint-disable max-len */
import React from 'react';
import { render, cleanup, fireEvent, waitFor, screen } from '@testing-library/react';
import App from './App';

describe('color picker change and record undo and redo ', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  // ------------------------------------------------

  it('changes the color using record', async() => {
    render(<App />);

    const colorPicker = await screen.findByLabelText('Current Color');
    const colorDisplay = await screen.findByTestId('frig');

    fireEvent.change(colorPicker, {
      target:{
        value: '#0000FF'
      }
    });
    
    return waitFor(() => {
      expect(colorPicker).toHaveValue('#0000ff');
      expect(colorDisplay.style.backgroundColor).toBe('rgb(0, 0, 255)');
    });

  });

  // ------------------------------------------------


  it('undoes the color ', async() => {
    render(<App />);

    const colorPicker = await screen.findByLabelText('Current Color');
    const colorDisplay = await screen.findByTestId('frig');
    const colorUndo = await screen.findByText('undo');


    fireEvent.change(colorPicker, {
      target:{
        value: '#00FF00'
      }
    });

    fireEvent.click(colorUndo);
    
    return waitFor(() => {
      // expect(colorPicker).toHaveValue('#ff0000');
      expect(colorDisplay.style.backgroundColor).toBe('rgb(255, 0, 0)');
    });

  });

  // ------------------------------------------------

  it('redoes the color ', async() => {
    render(<App />);

    const colorPicker = await screen.findByLabelText('Current Color');
    const colorDisplay = await screen.findByTestId('frig');
    const colorUndo = await screen.findByText('undo');
    const colorRedo = await screen.findByText('redo');

    fireEvent.change(colorPicker, {
      target:{
        value: '#00FF00'
      }
    });

    fireEvent.click(colorUndo);
    fireEvent.click(colorRedo);

    return waitFor(() => {
      // expect(colorPicker).toHaveValue('#ff0000');
      expect(colorDisplay.style.backgroundColor).toBe('rgb(0, 255, 0)');
    });

  });


});
