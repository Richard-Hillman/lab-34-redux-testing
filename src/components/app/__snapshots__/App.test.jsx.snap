// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color picker change and record undo and redo  renders App 1`] = `
<DocumentFragment>
  <label
    for="COLOR_CHANGE"
  >
    Current Color
  </label>
  <input
    id="COLOR_CHANGE"
    type="color"
    value="red"
  />
  <div
    data-testid="frig"
    style="background-color: red; width: 10rem; height: 10rem;"
  />
</DocumentFragment>
`;
